DataBiosphere/Saturn-UI
# [TOAZ-15] [TUIM-1] Integrate new OAuth library to Terra UI to support B2C

### JIRAs:
* https://broadworkbench.atlassian.net/browse/TOAZ-15
* https://broadworkbench.atlassian.net/browse/TUIM-1

### Description
This is an un-revert of [the original PR](https://github.com/DataBiosphere/terra-ui/pull/2949) to swap out Terra UI's oauth library to support Google and B2C. The original change was reverted because it was causing session timeouts to happen too frequently. The reason this started happening is the `gapi` client library is able to keep access tokens fresh indefinitely through some magic -- they render an iframe on a Google server and are able to store/reload previous auth responses to renew expired access tokens. The new library wasn't doing that, so access tokens would expire if the browser wasn't able to renew them in time (e.g. if the computer was asleep). 

The OAuth solution for this is refresh tokens. We now request an access token _and_ a refresh token from Google (or B2C). The refresh token can be used to renew an expired Google access token. Terra UI `signOut` clears the access token and refresh token from the browser, so we still adhere to the session timeout requirement for sensitive users. TDR has implemented a similar change here: https://github.com/DataBiosphere/jade-data-repo-ui/pull/1132. @dinvlad is in the loop on this from an AppSec/compliance perspective (Denis feel free to correct anything here).

### Commits:
* The [first commit](https://github.com/DataBiosphere/terra-ui/pull/3117/commits/5649d1615034a7cc65c633f814141b9ffb441120) is a pure un-revert of the original change.
* The [second commit](https://github.com/DataBiosphere/terra-ui/pull/3117/commits/3dbd5c92b26b6d0126d1e9974b621a806433597f) adds functionality to request and use a refresh token

### Testing notes:
1. Test on PR site if possible. While developing I observed different behavior on the PR site vs localhost w.r.t. refresh tokens.
2. Verify user login and access to all Terra UI screens.
3. Verify the billing page in particular -- it should automatically elevate scopes and display billing accounts if the user has previously granted `cloud-billing` to the application; otherwise it should display the Google prompt to elevate scopes (which is actually blocked on non-prod environments).
4. Verify session timeout. This is kind of annoying to test. I've tried leaving Terra UI open all day while I was working; and also all night while my laptop was asleep. In both cases I was still signed in when I came back, which is the same behavior as Terra prod today.

Thank you!
